src/main/java/su/terrafirmagreg/core/common/data/machines/TFGMachines.java

```diff
@@ -34,6 +64,66 @@ public static void init() { }
 //				.renderer(() -> new WorkableSteamMachineRenderer(pressure, GTCEu.id("block/machines/aqueous_accumulator")))
 //				.register());
 
+	public static final MultiblockMachineDefinition ELECTRIC_GREENHOUSE = REGISTRATE.multiblock("electric_greenhouse", GreenhouseMachine::new)
+		.rotationState(RotationState.NON_Y_AXIS)
+		.recipeType(TFGRecipeTypes.GREENHOUSE_RECIPES)
+		.recipeModifier(GTRecipeModifiers.OC_PERFECT)
+		.appearanceBlock(GTBlocks.CASING_STEEL_SOLID)
+		.workableCasingRenderer(GTCEu.id("block/casings/solid/machine_casing_solid_steel"), GTCEu.id("block/multiblock/implosion_compressor"), false)
+		.pattern(definition -> {
+			TraceabilityPredicate energyHatch = Predicates.abilities(PartAbility.INPUT_ENERGY).setMinGlobalLimited(1).setMaxGlobalLimited(2);
+			TraceabilityPredicate importBus = Predicates.abilities(PartAbility.IMPORT_ITEMS).setExactLimit(1);
+			TraceabilityPredicate exportBus = Predicates.abilities(PartAbility.EXPORT_ITEMS).setExactLimit(1);
+			TraceabilityPredicate importHatch = Predicates.abilities(PartAbility.IMPORT_FLUIDS).setExactLimit(1);
+			TraceabilityPredicate maintHatch = Predicates.autoAbilities(true, false, false).setExactLimit(1);
+
+			return FactoryBlockPattern.start()
+			.aisle("CCCCCCC", "XXXFXXX", "XXXFXXX", "XXXFXXX", "XXXFXXX", "XXXFXXX", "XXXFXXX", "XXXFXXX", "   F   ")
+			.aisle("CDDDDDC", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", " XXFXX ")
+			.aisle("CDDDDDC", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", " XXFXX ")
+			.aisle("CDDDDDC", "F#####F", "F#####F", "F#####F", "F#####F", "F#####F", "F#####F", "F#####F", "FFFFFFF")
+			.aisle("CDDDDDC", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", " XXFXX ")
+			.aisle("CDDDDDC", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", "X#####X", " XXFXX ")
+			.aisle("CCCYCCC", "XXXFXXX", "XXXFXXX", "XXXFXXX", "XXXFXXX", "XXXFXXX", "XXXFXXX", "XXXFXXX", "   F   ")
+			.where('C', Predicates.blocks(GTBlocks.CASING_STEEL_SOLID.get()).or(importBus).or(exportBus).or(importHatch).or(maintHatch).or(energyHatch))
+			.where('Y', Predicates.controller(Predicates.blocks(definition.get())))
+			.where('#', Predicates.air())
```

Comment:
we'll want to change this back to allowing logs and leaf blocks, as well as the preview structure, but we can do that later